source/extensions/filters/http/file_system_buffer/filter.cc

```diff
@@ -7,12 +7,459 @@ namespace FileSystemBuffer {
+  if (config_) {
+    return true;
+  }
+  auto route = request_callbacks_->route();
+  FilterConfigVector config_chain;
+  if (route) {
+    route->traversePerFilterConfig(
+        FileSystemBufferFilter::filterName(),
+        [&config_chain](const Router::RouteSpecificFilterConfig& route_cfg) {
+          auto cfg = dynamic_cast<const FileSystemBufferFilterConfig*>(&route_cfg);
+          config_chain.emplace_back(*cfg);
+        });
+  }
+  config_chain.emplace_back(*base_config_);
+  config_.emplace(config_chain);
+  request_state_.setConfig(config_->request());
+  response_state_.setConfig(config_->response());
+  return config_->hasAsyncFileManager();
+}
+
+Http::FilterHeadersStatus FileSystemBufferFilter::onBadConfig() {
+  filterError("config was missing manager_config");
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterHeadersStatus FileSystemBufferFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                                bool end_stream) {
+  // If there is no body then we're done here.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  if (!initPerRouteConfig()) {
+    return onBadConfig();
+  }
+  if (request_state_.config_->behavior().bypass()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  // If there's already a content length header we don't need to inject one unless the configuration
+  // says specifically to overwrite any incoming content-length.
+  request_state_.injecting_content_length_header_ =
+      (request_state_.config_->behavior().injectContentLength() &&
+       headers.ContentLength() == nullptr) ||
+      request_state_.config_->behavior().replaceContentLength();
+  // This effectively limits the output buffer to the size of the memory buffer, which should keep
+  // us from just draining all the streaming data into that 'hidden' memory buffer. It unfortunately
+  // means we still have the thread potentially using twice the memory that the user configured as
+  // the memory limit, once in our own buffer and once in the outgoing buffer. (Plus overflow
+  // because the limit isn't hard.)
+  request_callbacks_->setDecoderBufferLimit(request_state_.config_->memoryBufferBytesLimit());
+  // If we're going to buffer everything, don't even start sending the data until we're ready.
+  if (request_state_.injecting_content_length_header_ ||
+      request_state_.config_->behavior().alwaysFullyBuffer()) {
+    request_headers_ = &headers;
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus FileSystemBufferFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                                bool end_stream) {
+  // If there is no body then we're done here.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  if (!initPerRouteConfig()) {
+    return onBadConfig();
+  }
```

Comment:
Seems like we should only do this once in decodeHeaders() and then remember the failure?

Reply:
It's possible to get through decodeHeaders without having called initPerRouteConfig, if there was no body in the request. And then it's possible, if the request was HEAD so the response body is also empty, to *never* need to call initPerRouteConfig.

If there was a failure the first time then we never even reach the second one because we already did sendLocalReply so encodeHeaders is not reached, and if there was a success the first time then the second one just uses the already-determined value. (Unless I misunderstand something.)